README.md
README-LakeMonitoR
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

    #> Last Update: 2021-07-13 09:18:50

# LakeMonitoR

Suite of functions and tools for lake monitoring.

# Badges

[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://GitHub.com/leppott/LakeMonitoR/graphs/commit-activity)
[![lifecycle](https://img.shields.io/badge/lifecycle-stable-green.svg)](https://www.tidyverse.org/lifecycle/#stable)

[![License:
GPL-3](https://img.shields.io/badge/license-GPL--3-blue.svg)](https://cran.r-project.org/web/licenses/GPL-3)

[![CodeFactor](https://www.codefactor.io/repository/github/leppott/LakeMonitoR/badge)](https://www.codefactor.io/repository/github/leppott/LakeMonitoR)
[![codecov](https://codecov.io/gh/leppott/LakeMonitoR/branch/main/graph/badge.svg?token=C2MPX70BAL)](https://codecov.io/gh/leppott/LakeMonitoR)
[![R-CMD-check](https://github.com/leppott/LakeMonitoR/workflows/R-CMD-check/badge.svg)](https://github.com/leppott/LakeMonitoR/actions)

[![GitHub
issues](https://img.shields.io/github/issues/leppott/LakeMonitoR.svg)](https://GitHub.com/leppott/LakeMonitoR/issues/)

[![GitHub
release](https://img.shields.io/github/release/leppott/LakeMonitoR.svg)](https://GitHub.com/leppott/LakeMonitoR/releases/)
[![Github all
releases](https://img.shields.io/github/downloads/leppott/LakeMonitoR/total.svg)](https://GitHub.com/leppott/LakeMonitoR/releases/)

# Installation

To install the current version of the code from GitHub use the example
below.

``` r
if(!require(remotes)){install.packages("remotes")}  #install if needed
remotes::install_github("leppott/LakeMonitoR")
```

The vignette (big help file) isn’t created when installing from GitHub
with the above command. If you want the vignette download the compressed
file from GitHub and install from that file or install with the command
below. The “force = TRUE” command is used to ensure the package will
install over and existing install of the same version (e.g., the same
version without the vignettes).

``` r
if(!require(remotes)){install.packages("remotes")}  #install if needed
remotes::install_github("leppott/LakeMonitoR", force = TRUE, build_vignettes = TRUE)
```

# Purpose

Aid analysis of temperature and dissolved oxygen depth profiles from
lakes.

# Issues

<https://github.com/leppott/LakeMonitoR/issues>

# Documentation

Included Vignette and install guide.

# Shiny App

A Shiny app is included in the package.

The online version is hosted at
<https://tetratech-wtr-wne.shinyapps.io/LakeMonitoR>

The Shiny app can be run from R console using the shiny package without
installing the LakeMonitoR package.

``` r
if(!require(shiny)){install.packages("shiny")}
shiny::runGitHub(repo = "LakeMonitoR"
                , username = "leppott"
                , ref = "main"
                , subdir = "inst/shiny-examples/LakeMonitoR")
```


NEWS
NEWS
================
<Erik.Leppo@tetratech.com> and <Tim.Martin@state.mn.us>

<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->

    #> Last Update: 2023-02-10 14:12:23

# LakeMonitoR 1.0.0.9012

Release - 2023-02-10

- fix: Use `na.omit` for `tdox` calculation in Shiny, Issue \#24
  - Shiny would crash with too many mismatched DO and temperature
    measurements

# LakeMonitoR 1.0.0.9011

Release - 2022-11-14

- fix: Update handling of data to allow schmidt data and plot function
  to work

# LakeMonitoR 1.0.0.9010

Release - 2022-11-14

- refactor: Shiny heatmap display not working with all files as plotly
  - Change to static ggmap object

# LakeMonitoR 1.0.0.9009

Release - 2022-11-14

- fix: Update shiny display plots to work with non-standard dates

# LakeMonitoR 1.0.0.9008

Release - 2022-11-11

- fix: Update rLakeAnalyzer portion of Shiny app

# LakeMonitoR 1.0.0.9007

Release - 2022-11-04

- feat: Add function to aid in transforming date time fields
  - fun.DateTimeFormat
  - Update Vignette to show its use
  - Update Shiny app to use the new function on imported data

# LakeMonitoR 1.0.0.9006

Release - 2022-10-21

- fix: typo in plot_heatmap updates
- refactor: Change shiny to use rLA style heat map.

# LakeMonitoR 1.0.0.9005

Release - 2022-10-21

- refactor: Change heatmap
  - contours to FALSE
  - colors to rLakeAnalyzer colors

# LakeMonitoR 1.0.0.9004

Release - 2021-09-30

- refactor: Update Shiny calculate sidebar “No Date” to “No Depth”
- refactor: Update Shiny calculate sidebar to remove WSPD from label

# LakeMonitoR 1.0.0.9003

Release - 2021-08-24

- fix: Update tdox() from lm to glm for calculation

# LakeMonitoR 1.0.0.9002

Release - 2021-08-23

- fix: Update tdox() to use input value for DO instead of 3
- refactor: Update tdox() example for tdox_4

# LakeMonitoR 1.0.0.9001

Release - 2021-08-05

- fix: Update stratification() plot for julian day
  - %j starts with 1 so subtract 1 to keep Jan 1 as day 1 with origin
    date of “2004-01-01”.
- refactor: Update stratification() plot with range of data, Issue \#22
- refactor: Update stratification() plot with theme_bw() to match other
  plots

# LakeMonitoR 1.0.0

Release - 2021-08-02

- docs: Bump to version 1.0 for release

# LakeMonitoR 0.1.1.9016

Release - 2021-08-02

- fix: Update Shiny app folder for aggregate function
  - Needed the same fix as Calculation files

# LakeMonitoR 0.1.1.9015

Release - 2021-08-02

- fix: Update Shiny app order of operations for import/upload of files
  - Code was removing the Results/data-input folder and not recreating
    it
- refactor: Change default oxythermal temp from 30 to 20 in Shiny app

# LakeMonitoR 0.1.1.9014

Release - 2021-07-29

- refactor: Update Shiny Server.R
  - Rework directory exists code, make consistent
  - Update sink message for area file
  - Update file name for oxythermal plot output
  - Move DO plots inside of DO if statement with oxythermal and TDOx
  - Add error condition for oxythermal plot when no data to plot
    - For when have units (F) different than default threshold (30 C)
  - Change TDOx file output name
  - Update plots for measurement 2
    - heat map 2
    - depth profile 2

# LakeMonitoR 0.1.1.9013

Release - 2021-07-29

- refactor: Update Shiny rLA plot names to be consistent.
  - Use plot_rLA_x.png

# LakeMonitoR 0.1.1.9012

Release - 2021-07-29

- refactor: Update Shiny rLA plot names to be consistent.
  - Use plot_rLA_x.png

# LakeMonitoR 0.1.1.9011

Release - 2021-07-28

- refactor: Change default temp operator in plot_oxythermal
  - Update help text and function defaults to match example
- refactor: Update Shiny app
  - Calculation
    - Add Stratification method pull down
    - Add oxythermal inputs (temp and do)
    - Add TDOx value input
- function: Add Shiny app function to calculate stratification with
  different methods based on user input
- docs: Add tidyr package to DESCRIPTION

# LakeMonitoR 0.1.1.9010

Release - 2021-07-26

- docs: Update vignette examples

# LakeMonitoR 0.1.1.9009

Release - 2021-07-26

- docs: Rename vignette
- docs: Add plotting example for TDO3 to vignette, Issue \#21

# LakeMonitoR 0.1.1.9008

Release - 2021-07-16

- fix: Stratification, update column reference. Undo change in
  v0.1.1.9007
- feature: Add TDOx function, Issue \#21

# LakeMonitoR 0.1.1.9007

Release - 2021-07-16

- refactor: Stratification, update a hard coded column reference
- refactor: Stratification, update code to allow for only 2 depths,
  Issue \#20

# LakeMonitoR 0.1.1.9006

Release - 2021-07-16

- fix: Change Results folder reference in global to drop “.”
  - Fails in Linux with the “.” in file.path

# LakeMonitoR 0.1.1.9005

Release - 2021-07-16

- fix: Revert Shiny app to v0.1.1 and redo changes
  - Shiny app Help as HTML conflicts with file upload. Redesign as
    non-HTML
  - Remove semicolon as available delimiter for import files
  - Shiny Calculate tab labels ‘measurement’ to ‘value’
  - Ensure all uploaded data files going to Results/data_input
  - Add tooltips to tab_Calc
  - Add `ShinyBS` to global.R
- feat: Add missing function fun.write.log() to fun.Helper.R
- refactor: Rename log files with prefix “*log*”
- refactor: Use global “path_results” in server.R
- docs: Update Vignette with `plot_oxythermal()` function
- docs: Update oxythermal plots in function help and vignette
  - example to use \<= 68 rather than \>= 68
  - Combination plot; oxythermal, heat map temp, heat map DO

# LakeMonitoR 0.1.1.9004

Release - 2021-07-13

- fix: Unwind some changes to Shiny app. App won’t allow for upload of
  files
  - Combine files tab is now working.
- fix: `AggregateFiles()` to `agg_depth_files()` in server.R
- fix: Shiny, in Aggregate section replace “import” folder with
  “Results/data_input” folder

# LakeMonitoR 0.1.1.9003

Release - 2021-07-13

- refactor: Shiny app update.
- docs: Update README with example to run Shiny app.

# LakeMonitoR 0.1.1.9002

Release - 2021-07-09


R/tdox.R
#' @title Temperature Dissolved Oxygen x (TDOx)
#'
#' @description Calculates Temperature at Dissolved Oxygen Level X
#'
#' @details Using depth profiles for both temperature and dissolved oxygen
#' the temperature value at x DO is calculated.
#'
#' Values for min, mean, and max are calculated for daily and annual values.
#'
#' The calculation is the temperature at the depth where the value x of DO
#' occurs.  This is done by interpolating the DO value of x and associating the
#' temperature.
#'
#' The calculation is based on information in:
#'
#' P.C. Jacobson, H.G. Stefan, and D.L. Pereira.  2010. Coldwater fish
#' oxythermal habitat in Minnesota lakes: influence of total phosphorus, July
#' air temperature, and relative depth.  Canadian Journal of Fisheries and
#' Aquatic Sciences 67(12):2002-2013.  DOI 10.1139/F10-115
#'
#' The calculation is achieved in R by using the function `glm(temp ~ do)` for
#' each temperature and DO profile at each date and time.  Then `predict.glm()`
#' is used to generate the temperature at a DO value of x.
#'
#' From these calculations min, mean, and max is generated for each day and year
#' present in the data.
#'
#' @param data data frame
#' @param col_date Column name, Date
#' @param col_depth Column name, Depth
#' @param col_temp Column name, temperature
#' @param col_do Column name, Dissolved Oxygen
#' @param do_x_val DO value from which to generate temperature.  Default = 3
#'
#' @return A data frame with min, mean, and max for each year and day.
#'
#' @examples
#' ## Example 1, entire data set
#'
#' # data
#' data <- laketest
#'
#' # Columns
#' col_date  <- "Date.Time"
#' col_depth <- "Depth"
#' col_temp  <- "temp_F"
#' col_do    <- "DO_conc"
#' do_x_val  <- 3
#'
#' tdox_3 <- tdox(data = data
#'                , col_date = col_date
#'                , col_depth = col_depth
#'                , col_temp = col_temp
#'                , col_do = col_do
#'                , do_x_val = do_x_val)
#'
#' tdox_3
#'
#'
#' #~~~~~~~~~~~~~~~~~~~~
#'
#' ## Example 2, small dataset and different DO value
#'
#' do <- c(8.2, 8.2, 8.2, 8.2, 8, 7.5, 5.75, 4.4, 2.5)
#' temp <- c(26, 26, 26, 26, 25, 24.9, 24.8, 24.25, 23.5)
#' depth <- c(0.5, 1, 2, 3, 4, 5, 5.5, 6, 6.5)
#' date <- Sys.time()
#' data2 <- data.frame("datetime" = date
#'                    , "depth" = depth
#'                    , "temp" = temp
#'                    , "do" = do)
#'
#' tdox_4 <- tdox(data = data2
#'                , col_date = "datetime"
#'                , col_depth = "depth"
#'                , col_temp = "temp"
#'                , col_do = "do"
#'                , do_x_val = 4)
#'
#' tdox_4
#'
#' #~~~~~~~~~~~~~~~~~~~~
#'
#' ## Example 3, Single profile with plot
#'
#' # Data
#'data <- laketest
#'
#' # Columns
#' col_date  <- "Date.Time"
#' col_depth <- "Depth"
#' col_temp  <- "temp_F"
#' col_do    <- "DO_conc"
#' do_x_val  <- 3
#'
#' df_plot <- data[data[, "Date.Time"] == "2018-07-03 12:00", ]
#'
#' tdox_3_plot <- tdox(data = df_plot
#'                     , col_date = col_date
#'                     , col_depth = col_depth
#'                     , col_temp = col_temp
#'                     , col_do = col_do
#'                     , do_x_val = do_x_val)
#'
#' tdox_3_plot
#'
#' # Plot
#' ## Temp, x-bottom, circles
#' plot(df_plot$temp_F, df_plot$Depth, col = "black"
#'      , ylim = rev(range(df_plot$Depth))
#'      , xlab = "Temperature (F)", ylab = "Depth (m)")
#' lines(df_plot$temp_F, df_plot$Depth, col = "black")
#' abline(v = 53.66, col = "black", lty = 3, lwd = 2)
#' ## plot on top of existing plot
#' par(new=TRUE)
#' ## Add DO
#' plot(df_plot$DO_conc, df_plot$Depth, col = "blue", pch = 17
#'      , ylim = rev(range(df_plot$Depth))
#'      , xaxt = "n", yaxt = "n", xlab = NA, ylab = NA)
#' lines(df_plot$DO_conc, df_plot$Depth, col = "blue")
#' axis(side = 3, col.axis = "blue", col.lab = "blue")
#' mtext("Dissolved Oxygen (mg/L)", side = 3, line = 3, col = "blue")
#' abline(v = 3, col = "blue", lty = 3, lwd = 2)
#' ## Legend
#' legend("bottomright"
#'        , c("Temperature", "Temp = 53.66 @ DO = 3", "Dissolved Oxygen", "DO = 3")
#'        , col = c("black", "black", "blue", "blue")
#'        , pch = c(21, NA, 17, NA)
#'        , lty = c(1, 3, 1, 3)
#'        , lwd = c(1, 2, 1, 2))
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @export
tdox <- function(data
                 , col_date
                 , col_depth
                 , col_temp
                 , col_do
                 , do_x_val = 3) {

  # global variable bindings


  # Ensure proper formats
  # Convert to data frame (avoid use of tibble)
  data <- data.frame(data)
  data[, col_date]  <- as.Date(data[, col_date])
  data[, col_depth] <- as.numeric(data[,col_depth])
  data[, col_temp]  <- as.numeric(data[, col_temp])
  data[, col_do]    <- as.numeric(data[, col_do])

  # Dates
  Dates <- sort(unique(data[, col_date]))
  n_Dates <- length(Dates)

  # QC days to numeric
  n_Dates <- as.numeric(n_Dates)


  # for calculations
  # StratificationAnalysis <- NULL
  # meandata <- NULL

  # Calc, Predicted Value ----
  df_tdox_pred <- NULL
  #
  for (i in Dates){

    # filter for date
    calcdate <- data[data[, col_date] == i, ]
    # filter for NA
    #calcdate <- calcdate[!is.na(calcdate[, col_measure]), ]
    # Define DO and Temp for LM
    temp <- calcdate[, col_temp]
    do <- calcdate[, col_do]
    # Calc LM
    lm_tdox <- glm(temp ~ do)
    # Predict Value
    predict_tdox <- max(predict.glm(lm_tdox, data.frame("do" = do_x_val))
                        , na.rm = TRUE)


    singledate <- unique(calcdate[, col_date])
    stratrow <- data.frame("DateTime" = singledate, "TDOx" = predict_tdox)
    df_tdox_pred <- rbind(df_tdox_pred, stratrow)

  }## FOR ~ i ~ END
  df_tdox_pred <- df_tdox_pred[order(df_tdox_pred[, "DateTime"]), ]

  # Munge ----
  # Munge, Add Date and Year
  df_tdox_pred[, "Date"] <- format(df_tdox_pred[, "DateTime"], "%Y-%m-%d")
  df_tdox_pred[, "Year"] <- format(df_tdox_pred[, "DateTime"], "%Y")

  # Calc Stats, Year
  stats_year <- dplyr::summarise(dplyr::group_by(df_tdox_pred, Year)
                                 # Calculations
                                 , min = min(TDOx, na.rm = TRUE)
                                 , mean = mean(TDOx, na.rm = TRUE)
                                 , max = max(TDOx, na.rm = TRUE)
  )## summarise ~ Year ~ END
  #
  # Calc Stats, Date
  stats_date <- dplyr::summarise(dplyr::group_by(df_tdox_pred, Date)
                                 # Calculations
                                 , min = min(TDOx, na.rm = TRUE)
                                 , mean = mean(TDOx, na.rm = TRUE)
                                 , max = max(TDOx, na.rm = TRUE)
  )## summarise ~ Date ~ END


  # Combine Files
  names(stats_year)[1] <- "TimeFrame_Value"
  names(stats_date)[1] <- "TimeFrame_Value"
  stats_year[, "TimeFrame_Name"] <- "Year"
  stats_date[, "TimeFrame_Name"] <- "Date"

  stats_comb <- rbind(stats_year, stats_date)
  stats_comb[, "TDO_x_value"] <- do_x_val

  stats_comb <- stats_comb[, c("TimeFrame_Name"
                               , "TimeFrame_Value"
                               , "TDO_x_value"
                               , "min"
                               , "mean"
                               , "max")]


  # Results ----
  return(stats_comb)

}##FUNCTION ~ END

